peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+
+``ReadOnly`` can be used with ``ClassVar`` and ``Annotated`` in any nesting order:
+
+.. code-block:: python
+
+    class Foo:
+        foo: ClassVar[ReadOnly[str]] = "foo"
+        bar: Annotated[ReadOnly[int], Gt(0)]
+
+.. code-block:: python
+
+    class Foo:
+        foo: ReadOnly[ClassVar[str]] = "foo"
+        bar: ReadOnly[Annotated[int, Gt(0)]]
+
+This is consistent with the interaction of ``ReadOnly`` and :class:`typing.TypedDict`
+defined in :pep:`705`.
+
+``ClassVar`` excludes read-only attributes from being assignable to within
+initialization methods.
+
+Rules of ``Final`` should take priority when combined with ``ReadOnly``. As such,
+type checkers may warn on the redundancy of combining the two type qualifiers.
+
+
+Backwards Compatibility
+=======================
+
+This PEP introduces new contexts where ``ReadOnly`` is valid. Programs inspecting
+those places will have to change to support it. This is expected to mainly affect type checkers.
+
+However, caution is advised while using the backported ``typing_extensions.ReadOnly``
+in older versions of Python. Mechanisms inspecting annotations may behave incorrectly
+when encountering ``ReadOnly``; in particular, the ``@dataclass`` decorator
+which `looks for <https://docs.python.org/3/library/dataclasses.html#class-variables>`_
+``ClassVar`` will incorrectly treat ``ReadOnly[ClassVar[...]]`` as an instance attribute.
+
+
+Security Implications
+=====================
+
+There are no known security consequences arising from this PEP.
+
+
+How to Teach This
+=================
+
+[How to teach users, new and experienced, how to apply the PEP to their work.]
+
+
+Open Issues
+===========
+
+Assignment in ``__new__``
+-------------------------
+
+Immutable classes like :class:`fractions.Fraction` often do not define ``__init__``;
+instead, they perform initialization in ``__new__`` or classmethods. The proposed
+feature won't be useful to them. 
+
+OTOH, allowing assignment within ``__new__`` (and/or classmethods) could open way
+to non-trivial bugs:
+
+.. code-block:: python
+
+    class Foo:
+        # fully initialized objects
+        object_cache: ReadOnly[ClassVar[dict[int, Self]]] = {}
+
+        foo: ReadOnly[int]
+
+        def __new__(cls, foo: int) -> Self:
+            if foo + 1 in cls.object_cache:
+                # this instance is already initialized
+                self = cls.object_cache[foo + 1]
+
+            else:
+                # this instance is not
+                self = super().__new__(cls)
+
+            # assignment to an object which has been initialized before,
+            # breaking the invariant a read-only attribute can be assigned to
+            # only during its initialization?
+            self.foo = foo
+
+            cls.object_cache[foo] = self
+            return self
+
+To my understanding, properly detecting this problem would require type checkers
+to keep track of the "level of initialization" of an object.
```

Comment:
As discussed on Discourse, I think in practice this could be simplified to "was the object directly acquired from a call to super-type `__new__`", which I think is reasonable to track.

And it could also be an option to allow type checker to accept this unsoundness, in exchange for a simpler implementation.